File: library/src/main/java/su/levenetc/android/textsurface/animations/Alpha.java
package su.levenetc.android.textsurface.animations;

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.support.annotation.Nullable;

import su.levenetc.android.textsurface.utils.AnimatorEndListener;
import su.levenetc.android.textsurface.Text;
import su.levenetc.android.textsurface.interfaces.IEndListener;

/**
 * Created by Eugene Levenetc.
 */
public class Alpha extends AbstractSurfaceAnimation {

	private int to;
	private int from;
	private ValueAnimator animator;

	public static Alpha hide(Text text, int duration) {
		return new Alpha(text, duration, 255, 0);
	}

	public static Alpha show(Text text, int duration) {
		return new Alpha(text, duration, 0, 255);
	}

	public Alpha(Text text, int duration, int from, int to) {
		super(text, duration);
		this.from = from;
		this.to = to;
	}

	@Override public void start(@Nullable final IEndListener listener) {
		animator = ValueAnimator.ofInt(from, to);
		animator.setDuration(duration);
		animator.addUpdateListener(this);

		animator.addListener(new AnimatorEndListener() {
			@Override public void onAnimationEnd(Animator animation) {
				if (listener != null) listener.onAnimationEnd(Alpha.this);
			}
		});

		animator.start();
	}

	@Override public void onAnimationUpdate(ValueAnimator animation) {
		super.onAnimationUpdate(animation);
		text.setAlpha((int) animation.getAnimatedValue());
	}

	@Override public void cancel() {
		if (animator != null && animator.isRunning()) {
			animator.cancel();
			animator = null;
		}
	}
}

File: library/src/main/java/su/levenetc/android/textsurface/animations/Alpha.kt
package su.levenetc.android.textsurface.animations

import android.animation.Animator
import android.animation.ValueAnimator
import su.levenetc.android.textsurface.Text
import su.levenetc.android.textsurface.TextSurface
import su.levenetc.android.textsurface.interfaces.IEndListener
import su.levenetc.android.textsurface.utils.AnimatorEndListener

/**
 * Created by Eugene Levenetc.
 */
class Alpha(text: Text, duration: Long, private val from: Int, private val to: Int) : AbstractSurfaceAnimation(text, duration) {

    private var animator: ValueAnimator? = null
    override lateinit var textSurface: TextSurface

    override fun start(listener: IEndListener) {
        animator = ValueAnimator.ofInt(from, to)
        animator?.duration = duration
        animator?.addUpdateListener(this)
        animator?.addListener(object : AnimatorEndListener() {
            override fun onAnimationEnd(animation: Animator) {
                listener.onAnimationEnd(this@Alpha)
            }
        })
        animator?.start()
    }

    override fun onAnimationUpdate(animation: ValueAnimator) {
        super.onAnimationUpdate(animation)
        text.setAlpha(animation.animatedValue as Int)
    }

    override fun cancel() {
        if (animator != null && animator!!.isRunning) {
            animator!!.cancel()
            animator = null
        }
    }

    companion object {
        @kotlin.jvm.JvmStatic
        fun hide(text: Text, duration: Long): Alpha {
            return Alpha(text, duration, 255, 0)
        }

        @kotlin.jvm.JvmStatic
        fun show(text: Text, duration: Long): Alpha {
            return Alpha(text, duration, 0, 255)
        }
    }
}